linalg/amgxsolver.cpp

```diff
@@ -0,0 +1,1026 @@
+   {
+      Adisp[0] = 0;
+      for (int i=1; i<mpiTeamSz; ++i)
+      {
+         Adisp[i] = Adisp[i-1] + Apart[i-1];
+      }
+   }
+
+   MPI_Gatherv(inArr.HostRead(), inArr.Size(), MPI_INT,
+               outArr.HostWrite(), Apart.HostRead(), Adisp.HostRead(),
+               MPI_INT, 0, mpiTeamComm);
+}
+
+
+void AmgXSolver::GatherArray(const Array<int64_t> &inArr,
+                             Array<int64_t> &outArr,
+                             const int mpiTeamSz, const MPI_Comm &mpiTeamComm) const
+{
+   //Calculate number of elements to be collected from each process
+   Array<int> Apart(mpiTeamSz);
+   int locAsz = inArr.Size();
+   MPI_Gather(&locAsz, 1, MPI_INT,
+              Apart.GetData(),1, MPI_INT,0,mpiTeamComm);
+
+   MPI_Barrier(mpiTeamComm);
+
+   //Determine stride for process
+   Array<int> Adisp(mpiTeamSz);
+   int myid; MPI_Comm_rank(mpiTeamComm, &myid);
+   if (myid == 0)
+   {
+      Adisp[0] = 0;
+      for (int i=1; i<mpiTeamSz; ++i)
+      {
+         Adisp[i] = Adisp[i-1] + Apart[i-1];
+      }
+   }
+
+   MPI_Gatherv(inArr.HostRead(), inArr.Size(), MPI_INT64_T,
+               outArr.HostWrite(), Apart.HostRead(), Adisp.HostRead(),
+               MPI_INT64_T, 0, mpiTeamComm);
+
+   MPI_Barrier(mpiTeamComm);
+}
+
+void AmgXSolver::GatherArray(const Vector &inArr, Vector &outArr,
+                             const int mpiTeamSz, const MPI_Comm &mpiTeamComm,
+                             Array<int> &Apart, Array<int> &Adisp) const
+{
+   //Calculate number of elements to be collected from each process
+   int locAsz = inArr.Size();
+   MPI_Allgather(&locAsz, 1, MPI_INT,
+                 Apart.HostWrite(),1, MPI_INT, mpiTeamComm);
+
+   MPI_Barrier(mpiTeamComm);
+
+   //Determine stride for process
+   Adisp[0] = 0;
+   for (int i=1; i<mpiTeamSz; ++i)
+   {
+      Adisp[i] = Adisp[i-1] + Apart[i-1];
+   }
+
+   MPI_Gatherv(inArr.HostRead(), inArr.Size(), MPI_DOUBLE,
+               outArr.HostWrite(), Apart.HostRead(), Adisp.HostRead(),
+               MPI_DOUBLE, 0, mpiTeamComm);
+}
+
+void AmgXSolver::ScatterArray(const Vector &inArr, Vector &outArr,
+                              const int mpiTeamSz, const MPI_Comm &mpiTeamComm,
+                              Array<int> &Apart, Array<int> &Adisp) const
+{
+   MPI_Scatterv(inArr.HostRead(),Apart.HostRead(),Adisp.HostRead(),
+                MPI_DOUBLE,outArr.HostWrite(),outArr.Size(),
+                MPI_DOUBLE, 0, mpiTeamComm);
+}
+#endif
+
+void AmgXSolver::SetMatrix(const SparseMatrix &in_A, const bool update_mat)
+{
+
+   if (update_mat == false)
+   {
+
+      AMGX_matrix_upload_all(AmgXA, in_A.Height(),
+                             in_A.NumNonZeroElems(),
+                             1, 1,
+                             in_A.ReadI(),
+                             in_A.ReadJ(),
+                             in_A.ReadData(), NULL);
+
+      AMGX_solver_setup(solver, AmgXA);
+      AMGX_vector_bind(AmgXP, AmgXA);
+      AMGX_vector_bind(AmgXRHS, AmgXA);
+
```

Comment:
As a very minor point, this file has a lot of these blank lines that are not typical of MFEM style.